testgen/src/tester.rs

```diff
@@ -1,77 +1,196 @@
-        match self.full_path(path) {
-            None => None,
-            Some(full_path) => match fs::read_to_string(&full_path) {
-                Ok(file) => Some(file),
-                Err(_) => None,
-            },
+    pub fn cleanup(&self) -> Option<()> {
+        fs::remove_dir_all(&self.current_dir)
+            .ok()
+            .and(fs::create_dir_all(&self.current_dir).ok())
+    }
+
+    pub fn push(&self, child: &str) -> Option<Self> {
+        let mut path = PathBuf::from(&self.current_dir);
+        path.push(child);
+        path.to_str().and_then(|path| TestEnv::new(path))
+    }
+
+    pub fn current_dir(&self) -> &str {
+        &self.current_dir
+    }
+
+    pub fn logln(&self, msg: &str) -> Option<()> {
+        println!("{}", msg);
+        self.full_path("_log").and_then(|full_path| {
+            fs::OpenOptions::new()
+                .create(true)
+                .append(true)
+                .open(full_path)
+                .ok()
+                .and_then(|mut file| file.write_all((String::from(msg) + "\n").as_bytes()).ok())
+        })
+    }
+
+    pub fn logln_to(&self, msg: &str, rel_path: &str) -> Option<()> {
+        println!("{}", msg);
+        self.full_path(rel_path).and_then(|full_path| {
+            fs::OpenOptions::new()
+                .create(true)
+                .append(true)
+                .open(full_path)
+                .ok()
+                .and_then(|mut file| file.write_all((String::from(msg) + "\n").as_bytes()).ok())
+        })
+    }
+
+    /// Read a file from a path relative to the environment current dir into a string
+    pub fn read_file(&self, rel_path: &str) -> Option<String> {
+        self.full_path(rel_path)
+            .and_then(|full_path| fs::read_to_string(&full_path).ok())
+    }
+
+    /// Write a file to a path relative to the environment current dir
+    pub fn write_file(&self, rel_path: &str, contents: &str) -> Option<()> {
+        self.full_path(rel_path)
+            .and_then(|full_path| fs::write(full_path, contents).ok())
+    }
+
+    /// Parse a file from a path relative to the environment current dir as the given type
+    pub fn parse_file_as<T: DeserializeOwned>(&self, rel_path: &str) -> Option<T> {
+        self.read_file(rel_path)
+            .and_then(|input| serde_json::from_str(&input).ok())
+    }
+
+    /// Copy a file from the path outside environment into the environment current dir
+    /// Returns the relative path of the file, or None if copying was not successful
+    pub fn copy_file_from(&self, path: &str) -> Option<String> {
+        let path = Path::new(path);
+        if !path.is_file() {
+            return None;
         }
+        path.file_name().and_then(|name| {
+            name.to_str().and_then(|name| {
+                self.full_path(name)
+                    .and_then(|dest| fs::copy(path, dest).ok().map(|_| name.to_string()))
+            })
+        })
```

Comment:
We could perhaps avoid the nesting here by using the `?` operator:

```suggestion
        let name = path.file_name()?.to_str()?;
        let dest = self.full_path(name)?;
        fs::copy(path, dest).ok()?;
        Some(name.to_string())
```

Reply:
yes, this works, nice:)